### PR TITLE
drivers: spi: nrfx_spi(m|s): attempt to use instance based instantiation

### PR DESCRIPTION
**Pull request by gmarull**
This PR is an attempt to showcase an option to make nRF drivers instance based. At the same time, some refactoring to improve readability/consistency has been done as well.

## Context

nRF drivers have never used instance based instantiation, mainly because of how HAL is designed (it provides some utilities that rely on the peripheral index, e.g. pre-defined IRQ callbacks). As a result, drivers tend to be more complex than others, for example, they still require per-instance Kconfig options. They also make use of certain HAL definitions by concatenating tokens, for example, to obtain the register addresses (instead of using the one from DT). This PR attempts to align the SPI driver with common practices in-tree, while still keeping nrfx HAL.

NOTE: I haven't tested all changes, this is just a draft that likely needs some more work.

### CODE REVIEW

File: drivers/spi/spi_nrfx_spim.c [hunk 1]
```diff
@@ -480,140 +507,128 @@ static int spim_nrfx_pm_action(const struct device *dev,
...
-		))							       \
-		err = spi_context_cs_configure_all(&dev_data->ctx);	       \
-		if (err < 0) {						       \
-			return err;					       \
-		}							       \
-		spi_context_unlock_unconditionally(&dev_data->ctx);	       \
-		COND_CODE_1(CONFIG_SOC_NRF52832_ALLOW_SPIM_DESPITE_PAN_58,     \
-		(return anomaly_58_workaround_init(dev);),		       \
-		(return 0;))						       \
+		IRQ_CONNECT(DT_INST_IRQN(inst), DT_INST_IRQ(inst, priority),   \
```
Comment (carlescufi) [inline drivers/spi/spi_nrfx_spim.c:586]: Even if we don't use everything in this PR, we need to keep this. We are currently ignoring the IRQ number in the .dts(i) files, and that is a problem. The same goes for `NRFX_SPIM_INSTANCE`.

**Review by carlescufi — COMMENTED** [review]
I really welcome this initiative to modernize and align with the rest of vendors. Also getting rid of Kconfig index-based options is a must, since those have no place in Zephyr anymore.

File: dts/arm/nordic/nrf51822.dtsi [hunk 2]
```diff
@@ -78,7 +79,9 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			reg = <0x40003000 0x1000>;
+			periph-idx = <0>;
```
Comment (carlescufi) [inline dts/arm/nordic/nrf51822.dtsi:82]: @mbolivar-nordic this is used by other vendors as well FYI.

File: modules/hal_nordic/nrfx/nrfx_config.h [hunk 3]
```diff
@@ -197,57 +198,80 @@
...
+#define SPI2_NODE DT_NODELABEL(spi2)
+#define SPI3_NODE DT_NODELABEL(spi3)
+#define SPI4_NODE DT_NODELABEL(spi4)
+
 #ifdef CONFIG_NRFX_SPI
 #define NRFX_SPI_ENABLED 1
-#endif
-#ifdef CONFIG_NRFX_SPI0
+
+#if DT_NODE_HAS_COMPAT_STATUS(SPI0, nordic_nrf_spi, okay)
```
Comment (anangl) [inline modules/hal_nordic/nrfx/nrfx_config.h:210]: This cannot be done this way. We need to keep the possibility of using nrfx drivers directly for some instances (the corresponding nodes in devicetree are disabled then) and Zephyr shims for others.
In my opinion we should keep options `NRFX_SPIx`, `NRFX_SPIMx`, and `NRFX_SPISx`, and select them in `drivers/spi/Kconfig.nrfx` when a given instance is enabled in devicetree with the corresponding compatible property.

**Review by anangl — CHANGES_REQUESTED** [review]

File: dts/arm/nordic/nrf51822.dtsi (hunk 2 above)
Comment (mbolivar-nordic) [reply dts/arm/nordic/nrf51822.dtsi:82]: > @mbolivar-nordic this is used by other vendors as well FYI.

There are 2 cases where this is used by other vendors:

1. Where the peripheral index is actually a value that is going to get written into a peripheral register
2. Where the author is stuffing software configuration into devicetree

Case 1. was explicitly discussed and approved, including by @galak. Case 2. is just laziness and catering to HAL restrictions in devicetree, and was explicitly disapproved by @galak. I have done my best to forbid additional cases, but some have crept in because I can't review everything in enough detail to enforce this policy.

In the case of the Nordic drivers, I am unaware of any register where we need this value, and so unless I am missing something it is clearly case 2, which is forbidden. I think the fix belongs in the HAL, or we should stop using the HAL.

File: drivers/spi/spi_nrfx_spim.c [hunk 4]
```diff
@@ -532,8 +532,7 @@ static int spim_nrfx_pm_action(const struct device *dev,
 	{								       \
 		struct spi_nrfx_data *dev_data = dev->data;		       \
 		int err;						       \
-		IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_SPIM##idx),		       \
-			    DT_IRQ(SPIM(idx), priority),		       \
+		IRQ_CONNECT(DT_IRQN(SPIM(idx)), DT_IRQ(SPIM(idx), priority),   \
```
Comment (mbolivar-nordic) [inline drivers/spi/spi_nrfx_spim.c]: ACK on this -- if you want to open a separate PR with this fix, we can merge it faster

File: drivers/spi/spi_nrfx_spim.c [hunk 5]
```diff
@@ -568,7 +568,7 @@ static int spim_nrfx_pm_action(const struct device *dev,
 	static const struct spi_nrfx_config spi_##idx##z_config = {	       \
 		.spim = NRFX_SPIM_INSTANCE(idx),			       \
 		.max_chunk_len = (1 << SPIM##idx##_EASYDMA_MAXCNT_SIZE) - 1,   \
-		.max_freq = SPIM##idx##_MAX_DATARATE * 1000000,		       \
+		.max_freq = SPIM_PROP(idx, max_frequency),		       \
```
Comment (mbolivar-nordic) [inline drivers/spi/spi_nrfx_spim.c]: That's fine, but perhaps we need some additional checks to make sure the user has not changed this value and that future SoC ports do not make any mistakes here.

I suggest adding some checks in soc/arm/nordic_nrf/, analogous to the validate_enabled_instances.c and validate_base_addresses.c files in that directory.

File: dts/bindings/spi/nordic,nrf-spi-common.yaml [hunk 6]
```diff
@@ -18,6 +18,14 @@ properties:
       description: |
         Maximum data rate the SPI peripheral can be driven at, in Hz.
 
+    overrun-character:
+      type: int
+      default: 0xff
+      description: |
+        The overrun character is used when all bytes from the TX buffer are
```
Comment (mbolivar-nordic) [inline dts/bindings/spi/nordic,nrf-spi-common.yaml:32]: This makes sense.

Suggestion to provide the user more hints about how this corresponds to registers:

```
        The overrun character (ORC) is used when all bytes from the TX buffer are
```

File: drivers/spi/spi_nrfx_spim.c [hunk 7]
```diff
@@ -26,6 +26,31 @@ LOG_MODULE_REGISTER(spi_nrfx_spim, CONFIG_SPI_LOG_LEVEL);
 #define SPI_BUFFER_IN_RAM 1
 #endif
 
+/* Maximum chunk length (depends on the EasyDMA bits, SoC dependent) */
```
Comment (mbolivar-nordic) [inline drivers/spi/spi_nrfx_spim.c:31]: Then let's put it in the SoC .dtsi, no? Why is this here instead of there?

File: drivers/spi/spi_nrfx_spis.c [hunk 8]
```diff
@@ -20,12 +20,39 @@ struct spi_nrfx_data {
 
 struct spi_nrfx_config {
 	nrfx_spis_t spis;
-	size_t      max_buf_len;
 #ifdef CONFIG_PINCTRL
 	const struct pinctrl_dev_config *pcfg;
 #endif
 };
 
+/* Maximum buffer length (depends on the EasyDMA bits, SoC dependent) */
```
Comment (mbolivar-nordic) [inline drivers/spi/spi_nrfx_spis.c:30]: Same comment throughout this PR

File: dts/bindings/spi/nordic,nrf-spim.yaml [hunk 9]
```diff
@@ -25,3 +25,28 @@ properties:
         Enables the workaround for the nRF52832 SoC SPIM PAN 58 anomaly.
         Must be used in conjunction with
         CONFIG_SOC_NRF52832_ALLOW_SPIM_DESPITE_PAN_58=y
+
+    has-rx-delay:
```
Comment (mbolivar-nordic) [inline dts/bindings/spi/nordic,nrf-spim.yaml:29]: Let's use `rx-delay-supported`, for consistency with things like `dfe-supported` in the [radio binding](https://docs.zephyrproject.org/latest/build/dts/api/bindings/net/wireless/nordic%2Cnrf-radio.html#dtbinding-nordic-nrf-radio).

File: dts/bindings/spi/nordic,nrf-spim.yaml [hunk 10]
```diff
@@ -25,3 +25,28 @@ properties:
...
+      type: boolean
+      required: false
+      description: |
+        Indicates if the SPIM instance has the capability of delaying MISO
+        sampling. This property needs to be defined at SoC level DTS files.
```
Comment (mbolivar-nordic) [inline dts/bindings/spi/nordic,nrf-spim.yaml:34]: Please align this description with the existing language in the radio binding, again for consistency.

File: dts/arm/nordic/nrf51822.dtsi (hunk 2 above)
Comment (mbolivar-nordic) [reply dts/arm/nordic/nrf51822.dtsi:82]: > unless I am missing something it is clearly case 2, which is forbidden.

I left this comment before beginning my review. Now that I got to the commit log, I have to NAK this approach on the grounds outlined above. I force other SoC vendors to play by these rules whenever we find people are trying; we should not make an exception for ourselves out of convenience here.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I get it, I really do, but I have to NAK certain aspects of this approach.

File: drivers/spi/spi_nrfx_spim.c [hunk 11]
```diff
@@ -480,140 +507,128 @@ static int spim_nrfx_pm_action(const struct device *dev,
...
-	IF_ENABLED(CONFIG_PINCTRL, (PINCTRL_DT_DEFINE(SPIM(idx))));	       \
-	static const struct spi_nrfx_config spi_##idx##z_config = {	       \
-		.spim = NRFX_SPIM_INSTANCE(idx),			       \
-		.max_chunk_len = (1 << SPIM##idx##_EASYDMA_MAXCNT_SIZE) - 1,   \
-		.max_freq = SPIM##idx##_MAX_DATARATE * 1000000,		       \
+									       \
+	static const struct spi_nrfx_config config##inst = {		       \
+		.spim = {						       \
+			.p_reg = (NRF_SPIM_Type *)DT_INST_REG_ADDR(inst),      \
+			.drv_inst_idx = DT_INST_PROP(inst, periph_idx),	       \
```
Comment (anangl) [inline drivers/spi/spi_nrfx_spim.c:612]: This is incorrect. The index is not a peripheral instance index, it is an nrfx_spim driver instance index (e.g. it will be 0, when only SPIM3 is enabled to be handled by the driver).

Comment (gmarull) [reply drivers/spi/spi_nrfx_spim.c:612]: good point, however, `NRFX_SPIM_INSTANCE` doesn't take device address from DT, which needs to be fixed. Overall, I think that in this case, HAL is making drivers more complicated because it overlaps with functionality covered by Zephyr. This is in fact a common pattern with HALs like nrfx (stateful), they tend to be inefficient and induce bad design patterns. We also have some other notable inefficiencies like this: https://github.com/zephyrproject-rtos/zephyr/blob/88ca3aca98c0788e2c45db5c3fb9c8523293a8c0/drivers/i2c/i2c_nrfx_twim.c#L295-L300,ie what could have been a simple register write replaced by a full deinit/init sequence because HAL doesn't support updating frequency.

**Comment by gmarull** [comment]
Closing this for now, will require changes to nrfx to do this, or, switch to nrf_/native drivers.
